### PR TITLE
feat: deploying scripts and predicates

### LINKED ISSUE #3262: Deploying predicates and scripts with blobs
[Duplicate of RS issue](https://github.com/FuelLabs/fuels-rs/issues/1519)

Like we currently do with contracts, we must be able to deploy script and predicate bytecode as blobs so that it can loaded in transactions to be executed more optimally.

### PR DESCRIPTION
**Pull request by arboleya**
- Closes #3262 

# Release notes

In this release, we:

- Added support for deploying scripts and predicates
- Upgraded to `fuel-core@0.37.0` 
- Upgraded to `forc@0.65.1`

# Checklist

- [x] All **changes** are **covered** by **tests** (or not applicable)
- [x] All **changes** are **documented** (or not applicable)
- [x] I **reviewed** the **entire PR** myself (preferably, on GH UI)
- [x] I **described** all **Breaking Changes** (or there's none)

### CODE REVIEW

**Comment by petertonysmith94** [comment]
@arboleya We changed the base to `next` to avoid having the `next` changes in this PR for now.

We will need to change this back to `master` once `next` has been merged.

File: .changeset/sharp-radios-fry.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+---
+
+feat: deploying scripts and predicates
```
Comment (petertonysmith94) [inline .changeset/sharp-radios-fry.md:4]: > [!NOTE]
> Change base branch to `master` once @arboleya has merged `next` into `master`.

File: packages/fuel-gauge/test/fixtures/forc-projects/Forc.toml [hunk 2]
```diff
@@ -8,6 +8,7 @@ members = [
   "auth_testing_contract",
   "bytecode-sway-lib",
   "bytes-contract",
+  "script-dummy",
```
Comment (petertonysmith94) [inline packages/fuel-gauge/test/fixtures/forc-projects/Forc.toml:11]: Use script that is already here.

File: packages/contract/src/contract-factory.ts [hunk 3]
```diff
@@ -372,6 +376,77 @@ export default class ContractFactory {
     return { waitForResult, contractId, waitForTransactionId };
   }
 
+  async deployAsBlobTxForScript(configurableConstants: { [name: string]: unknown } = {}): Promise<{
```
Comment (Dhaiwat10) [inline packages/contract/src/contract-factory.ts:379]: Why did we choose to add this logic on the contract factory and not on the script class?

File: packages/contract/src/loader/predicate-script-loader-instructions.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
+import { InstructionSet } from '@fuel-ts/program';
```
Comment (petertonysmith94) [inline packages/contract/src/loader/predicate-script-loader-instructions.ts:1]:
```suggestion
// Lifted from https://github.com/FuelLabs/fuels-rs/pull/1520/files#diff-ae7d8dde0236c6d422e1d30c8f24a21290ea9cb948cd70e67a301017ecb586a5R177
import { InstructionSet } from '@fuel-ts/program';
```

> [!NOTE]
> Change to actual branch once merged.

**Comment by petertonysmith94** [comment]
Spoken with @LuizAsFight, from a `fuel-connector` point of view and he does not foresee any breaking changes.
- Once we have a solution, we can update the `fuels` version.
- Update related `fuel-toolchain`
- Redeploy predicates
- Test connectors + wallets

File: packages/contract/src/contract-factory.ts (hunk 3 above)
Comment (petertonysmith94) [reply packages/contract/src/contract-factory.ts:379]: Were you thinking a script-factory or just on the script class itself?

File: packages/fuel-gauge/test/fixtures/forc-projects/Forc.toml (hunk 2 above)
Comment (petertonysmith94) [reply packages/fuel-gauge/test/fixtures/forc-projects/Forc.toml:11]: Going to resolve this as we can do a purge later.

File: .changeset/sharp-radios-fry.md (hunk 1 above)
Comment (petertonysmith94) [reply .changeset/sharp-radios-fry.md:4]: Resolved.

File: packages/fuels/src/cli/commands/deploy/deployPredicates.ts [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+ * Deploys all predicates.
+ */
+export async function deployPredicates(config: FuelsConfig) {
+  const predicates: DeployedPredicate[] = [];
+
+  const wallet = await createWallet(config.providerUrl, config.privateKey);
+
+  log(`Deploying predicates to: ${wallet.provider.url}`);
+
+  const predicatesLen = config.scripts.length;
```
Comment (luizstacio) [inline packages/fuels/src/cli/commands/deploy/deployPredicates.ts:53]:
```suggestion
  const predicatesLen = config.predicates.length;
```

File: packages/contract/src/contract-factory.ts [hunk 6]
```diff
@@ -372,6 +377,107 @@ export default class ContractFactory {
...
+    return Promise.resolve({
+      waitForResult: () =>
+        Promise.resolve({
+          transactionResult: {},
+          loaderBytecode: '',
+        }),
+      predicateRoot: '',
+      loaderBytecode: new Uint8Array(),
+      loaderBytecodeHexlified: '',
+    });
```
Comment (luizstacio) [inline packages/contract/src/contract-factory.ts:402]: Return offset to generate a abi copy for the loader with new offset of the data section.
```suggestion
     return Promise.resolve({
      waitForResult: () =>
        Promise.resolve({
          transactionResult: {},
          loaderBytecode: '',
          offset: 0,
        }),
      predicateRoot: '',
      loaderBytecode: new Uint8Array(),
      loaderBytecodeHexlified: '',
      offset: 0,
    });
```

File: .changeset/sharp-radios-fry.md [hunk 7]
```diff
@@ -0,0 +1,4 @@
+---
+---
```
Comment (petertonysmith94) [inline .changeset/sharp-radios-fry.md:2]:
```suggestion
---
"@fuel-ts/account": patch
"@fuel-ts/contract": patch
"@fuel-ts/fuels": patch
"@fuel-ts/program": patch
"@fuel-ts/versions": patch
"create-fuels": patch
---
```

File: .github/workflows/pr-release.yaml [hunk 8]
```diff
@@ -8,7 +8,7 @@ jobs:
     name: "Release PR to npm"
     runs-on: ubuntu-latest
     # comment out if:false to enable release PR to npm
-    if: false
+    # if: false
```
Comment (danielbate) [inline .github/workflows/pr-release.yaml:11]:
```suggestion
    if: false
```

File: .changeset/sharp-radios-fry.md [hunk 9]
```diff
@@ -0,0 +1,10 @@
...
+"@fuel-ts/account": patch
+"@fuel-ts/contract": patch
+"@fuel-ts/fuels": patch
```
Comment (danielbate) [inline .changeset/sharp-radios-fry.md:4]:
```suggestion
"fuels": patch
```

**Review by maschad — APPROVED** [review]

**Review by Torres-ssf — APPROVED** [review]

**Review by luizstacio — APPROVED** [review]

File: apps/docs-snippets/package.json [hunk 10]
```diff
@@ -4,8 +4,12 @@
   "description": "",
   "private": true,
   "scripts": {
-    "pretest": "run-s build:forc type:check",
-    "build:forc": "pnpm fuels build",
+    "pretest": "run-s fuels:build type:check",
```
Comment (danielbate) [inline apps/docs-snippets/package.json:7]: Will fix in #3263.

**Review by danielbate — APPROVED** [review]

**Pull request merged**